compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1203,11 +1206,36 @@ impl<'a> Parser<'a> {
                 if !this.recover_nested_adt_item(kw::Enum)? {
                     return Ok((None, TrailingToken::None));
                 }
+                let enum_field_start_span = this.token.span;
                 let ident = this.parse_field_ident("enum", vlo)?;
+                if this.token.kind == token::Colon {
+                    let snapshot = this.clone();
+                    this.bump();
+                    match this.parse_ty() {
+                        Ok(_) => {
+                            let mut err = this.struct_span_err(
+                                enum_field_start_span.to(this.prev_token.span),
+                                "the enum cannot have a struct field declaration",
+                            );
+                            err.span_suggestion_verbose(
+                                start_span,
+                                "consider using `struct` instead of `enum`",
```

Comment:
Hm, an `union` could also make sense here, wouldn't it? While it is comparatively much rarer, looking to define an `enum` like this may very well be thinking of an `union` in the first place. 

```suggestion
                                "consider defining a `struct` or an `union` instead of an `enum`",
```

Not entirely sure how or if `span_suggestion` can provide multiple alternative replacements, though.

Reply:
It might very well be that this suggestion is largely superfluous with the improved error message.